juju/juju
# State storage lifebinding

## Description of change

Drop the "binding" field and LifeBinding method for volumes and filesystems, and then drop the LifeBinder interface. Lifecycle binding was never user-specifiable, so we can drop it and simplify the rules for destroying storage.

Now, volumes and filesystems will always be destroyed along with the storage instances to which they are assigned. It is now only possible to have a volume unassigned to a storage instance by using "juju add-machine --disks", which only ever existed for testing purposes. Since the volumes you add in that way cannot be used with charms, we should have that flag ignored with a warning; this will prevent volumes being created that can only be destroyed by destroying the model.

## QA steps

SCENARIO 1
1. juju bootstrap aws
2. juju deploy postgresql --storage pgdata=ebs,10G
(wait for storage to be attached)
3. juju remove-application postgresql
(verify the storage and volume are removed)

SCENARIO 2
Same as above, but using rootfs pool.

## Documentation changes

None.

## Bug reference

None.
